Add complete Chatbot example with RAG capability for demo apps

WuhanMonkey opened the pull request:
# What does this PR do?

This complete example shows how to build a chatbot with RAG and streaming capability with Llama 2, using frameworks such as LangChain, FAISS, Gradio and Text-generation-inference. For RAG, we used Llama 2 Getting Started Guide published on Meta AI website as our knowledge reference. 

Fixes # (issue)

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ X] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [X ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After confirming the validity, we can start building the UI. Before we define the gradio blocks, let's first define the callback streams that we will use later for streaming feature.  \n",
```
A quick explanation of the gradio blocks?

streaming feature -> the streaming feature

WuhanMonkey replied:
Added the documentation link. Better explanation there.

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "metadata": {},
+   "source": [
+    "After confirming the validity, we can start building the UI. Before we define the gradio blocks, let's first define the callback streams that we will use later for streaming feature.  \n",
+    "This callback handler will put streaming LLM responses to a queue for gradio UI to render on the flight. "
```
on the flight -> on the fly

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "                model_selector = gr.Dropdown(list(model_dict.keys()), value=\"7b-chat\", label=\"Model\", info=\"Select the model\", interactive = True, scale=1)\n",
+    "                max_new_tokens_selector = gr.Number(value=512, precision=0, label=\"Max new tokens\", info=\"Adjust max_new_tokens\",interactive = True, minimum=1, maximum=1024, scale=1)\n",
+    "            with gr.Row():\n",
+    "                #hyperparameter selection\n",
```
hyperparameter selection -> temperature selection
as temperature is not really a hyperparameter.

WuhanMonkey replied:
Hmm, I think it is. At least in our context or habit. [Reference1](https://medium.com/@dixnjakindah/top-p-temperature-and-other-parameters-1a53d2f8d7d7#:~:text=There%20are%20many%20other%20hyperparameters,time%2C%20accuracy%2C%20and%20generalizability.). [Reference2](https://deepchecks.com/glossary/llm-parameters/#:~:text=One%20intriguing%20parameter%20within%20LLMs,of%20straying%20from%20the%20context.)

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "        else:\n",
+    "            return history + [[\"Invalid prompts - user prompt cannot be empty\", None]]\n",
+    "\n",
+    "    #chatbot logics for configuration, sending the prompts, rendering the streamed back genereations etc\n",
```
logics -> logic

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "    def input_cleanup():\n",
+    "        return \"\"\n",
+    "\n",
+    "    #when user click Enter and user message are submitted\n",
```
user click -> the user clicks
user message are -> the user message is

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "        bot, [model_selector, temperature_selector, top_p_selector, max_new_tokens_selector, user_prompt_message, chatbot, state], [chatbot, state]\n",
+    "    ).then(input_cleanup, [], [user_prompt_message], queue=False)\n",
+    "\n",
+    "    #when user click the submit button\n",
```
user click -> the user clicks

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "        bot, [model_selector, temperature_selector, top_p_selector, max_new_tokens_selector, user_prompt_message, chatbot, state], [chatbot, state]\n",
+    "    ).then(input_cleanup, [], [user_prompt_message], queue=False)\n",
+    "    \n",
+    "    #when user click the clear button\n",
```
user click -> the user clicks

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lastly, we can launch this demo on our localhost. Make sure you have access to the ports.  \n",
```
ports -> port

maybe a quick sentence on "how to make sure"?

WuhanMonkey replied:
Deleted this line to avoid confusion. Gradio will default launch on 7860 and if 7860 is not available it will launch on the next port available.

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "metadata": {},
+   "source": [
+    "Lastly, we can launch this demo on our localhost. Make sure you have access to the ports.  \n",
+    "In the notebook or a browser with URL http://0.0.0.0:7860 you should see the UI.  \n",
```
when can one open the URL http://0.0.0.0:7860? after running the following code?
demo.queue().launch(server_name="0.0.0.0")

also what's the port 7860? the default one?

A screenshot of the UI will be nice.

WuhanMonkey replied:
Rearragne the narrative to make it clear. 
I don't think a screenshot of UI is necessary though. You either launch it successfully or you can't reach anything in that port.

jeffxtang replied:
Just went through your whole notebook and was able to run the demo and see the Gradio UI both from the notebook and the browser with the URL. Maybe a better way than a screenshot, to make the demo more interesting  on a quick read (without having to run it), is to save the cell output like this:

![image](https://github.com/facebookresearch/llama-recipes/assets/535090/22c5937a-30d6-4897-9853-8f866b6c053f)

Three other comments:

1. About "Things to try in the chatbot demo:

Asking specific questions related to the Llama 2 Getting Started Guide
Streaming
Adjust parameters such as max new token generated"

Since "streaming" is not an element in the UI, is modifying it in the `llm = HuggingFaceTextGenInference(...)` call the only way to try it out? If so, maybe we don't list it here.

2. We probably should replace the code cell with

model = meta-llama/Llama-2-7b-chat-hf  
volume = $PWD/data  
token = #Your own HF tokens  
docker run --gpus all --shm-size 1g -e HUGGING_FACE_HUB_TOKEN=$token -p 8080:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:1.1.0 --model-id $model

with a Markdown cell and use ``` ... ``` in the cell, otherwise the user may think it's the same as other code cells in the notebook, run it and get an error (even though you mentioned "on a separate terminal".

3. The user may likely run the notebook on a separate machine than the machine running TGI, then using localhost will fail. So you may want to add a note asking them to replace "localhost" in the 3 cases with the IP visible to the machine running the notebook.
 
!curl localhost:8080...
LLAMA2_7B_HOSTPORT = "http://localhost:8080/"
LLAMA2_13B_HOSTPORT = "http://localhost:8080/" #Add your own host ports for model switching. You can host another TGI model on same instance on a different port.

WuhanMonkey replied:
Thanks @jeffxtang, accept all 4 comments/suggestions and updated.

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "Lastly, we can launch this demo on our localhost. Make sure you have access to the ports.  \n",
+    "In the notebook or a browser with URL http://0.0.0.0:7860 you should see the UI.  \n",
+    "Things to try in the chatbot demo:  \n",
+    "* Specific questions relates Llama 2 get started\n",
```
Specific questions relates Llama 2 get started -> Asking specific questions related to the Llama 2 Getting Started Guide

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "Things to try in the chatbot demo:  \n",
+    "* Specific questions relates Llama 2 get started\n",
+    "* Streaming\n",
+    "* Adjust hyperparameters such as max new token generated\n",
```
Adjust hyperparameters -> Adjusting parameters

jeffxtang commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "* Specific questions relates Llama 2 get started\n",
+    "* Streaming\n",
+    "* Adjust hyperparameters such as max new token generated\n",
+    "* Switch models with another container launched in a separate terminal"
```
Switch models -> Switching to another Llama model

jeffxtang reviewed (changes requested):
Hi Chester, thanks for all the work you've put into this! Most of my comments were suggestions about spellings/grammar. I'll try running the notebook and playing with the UI tomorrow and add comments/questions if any. 
@WuhanMonkey

WuhanMonkey commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+    "This notebook shows a complete example of how to build a Llama 2 chatbot hosted on your Browser that can answer questions based on your own data. We'll cover:\n",
+    "* The deployment process of Llama 2 7B with Text-generation-inference framework as a API server\n",
+    "* A chatbot example built with Gradio and wired to the server\n",
+    "* Add RAG capability with Llama 2 specific knowledge based on our Getting Started [guide](https://ai.meta.com/llama/get-started/)"
```
RAG is first mentioned in the title, so added the full name there. Thanks

WuhanMonkey commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Check the length and content of the doc to ensure we have loaded the right document with number of pages."
```
It actually has 37 pages

WuhanMonkey commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "source": [
+    "model = meta-llama/Llama-2-7b-chat-hf  \n",
+    "volume = $PWD/data  \n",
+    "token = Your own HF tokens  \n",
```
Not sure if you want me to delete this but changed the sentence as comment now

WuhanMonkey commented on demo_apps/RAG_Chatbot_example/RAG_Chatbot_Example.ipynb:
```diff
@@ -0,0 +1,631 @@
+   "outputs": [],
+   "source": [
+    "embeddings = HuggingFaceEmbeddings(model_name=\"sentence-transformers/all-MiniLM-L6-v2\",\n",
+    "                                       model_kwargs={'device': 'cpu'})\n",
```
I have added some comments before to talk about FAISS with GPU. I think I will stick with CPU as an example here.

WuhanMonkey commented:
Thank you @jeffxtang for the detailed review. I have addressed most of the comments and left some discussions open.

jeffxtang commented:
> Thank you @jeffxtang for the detailed review. I have addressed most of the comments and left some discussions open.

Thanks Chester for the update. Looks good! I added one hopefully final comment above, with 4 suggestions. @WuhanMonkey

WuhanMonkey commented:
> > Thank you @jeffxtang for the detailed review. I have addressed most of the comments and left some discussions open.
> 
> Thanks Chester for the update. Looks good! I added one hopefully final comment above, with 4 suggestions. @WuhanMonkey

Accepted the suggestions and updated!

albertodepaola commented:
On testing this, I noticed that the system_message does not get set unless the user clears the history. Additionally, the message_history is not being used in this sample, should we just remove it?

jeffxtang commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # Llama 2 Fine-tuning / Inference Recipes, Examples and Demo Apps
 
-**[Update Nov. 3, 2023] We recently released a series of Llama 2 demo apps [here](./demo_apps). These apps show how to run Llama 2 locally, in the cloud, and on-prem, and how to ask Llama 2 questions in general and about custom data (PDF, DB, or live).**
+**[Update Nov. 16, 2023] We recently released a series of Llama 2 demo apps [here](./demo_apps). These apps show how to run Llama 2 locally, in the cloud, on-prem, how to ask Llama 2 questions in general or about custom data (PDF, DB, or live) and end-to-end Chatbot implementation with RAG (Retrieval Augmented Generation).**
```
how to run Llama 2 locally, in the cloud, on-prem, how to ask Llama 2 questions in general or about custom data (PDF, DB, or live) and end-to-end Chatbot implementation
->
how to run Llama (locally, in the cloud, or on-prem), how to ask Llama questions in general or about custom data (PDF, DB, or live), how to integrate Llama with WhatsApp, and how to implement an end-to-end chatbot

WuhanMonkey replied:
Done, should be all set now!

HamidShojanazeri commented:
@WuhanMonkey can you pls disable the link check as @jeffxtang  did here, https://github.com/facebookresearch/llama-recipes/commit/52197ed6890eba83b5b8a289611797d8423216ed to pass CI.
